File: conductor/src/gates.ts
import type { ISdk } from 'iii-sdk'
import type { GateSpec } from './types.js'

export interface GateOutcome {
  ok: boolean
  reason?: string
}

export async function runGate(sdk: ISdk, gate: GateSpec, cwd: string): Promise<GateOutcome> {
  try {
    const result = await sdk.trigger<{ cwd: string }, GateOutcome | { ok: boolean; reason?: string }>({
      function_id: gate.function_id,
      payload: { cwd },
      timeoutMs: 600_000,
    })
    if (typeof result === 'object' && result !== null && 'ok' in result) {
      return { ok: Boolean(result.ok), reason: result.reason }
    }
    return { ok: false, reason: 'gate returned non-object' }
  } catch (err) {
    return { ok: false, reason: String(err) }
  }
}

export async function runAllGates(sdk: ISdk, gates: GateSpec[], cwd: string): Promise<Record<string, GateOutcome>> {
  const out: Record<string, GateOutcome> = {}
  for (const gate of gates) {
    out[gate.function_id] = await runGate(sdk, gate, cwd)
  }
  return out
}

export function allPassed(results: Record<string, GateOutcome>): boolean {
  return Object.values(results).every((r) => r.ok)
}

File: conductor/src/gates.rs
use std::path::Path;

use iii_sdk::{TriggerRequest, III};
use serde_json::json;

use crate::types::{GateOutcome, GateRunResult, GateSpec};

pub async fn run_gate(iii: &III, gate: &GateSpec, cwd: &Path) -> GateOutcome {
    let cwd_str = cwd.to_string_lossy().into_owned();
    let result = iii
        .trigger(TriggerRequest {
            function_id: gate.function_id.clone(),
            payload: json!({ "cwd": cwd_str }),
            action: None,
            timeout_ms: Some(600_000),
        })
        .await;
    match result {
        Ok(val) => match serde_json::from_value::<GateOutcome>(val.clone()) {
            Ok(o) => o,
            Err(_) => GateOutcome {
                ok: val
                    .get("ok")
                    .and_then(serde_json::Value::as_bool)
                    .unwrap_or(false),
                reason: val
                    .get("reason")
                    .and_then(serde_json::Value::as_str)
                    .map(String::from),
            },
        },
        Err(e) => GateOutcome {
            ok: false,
            reason: Some(e.to_string()),
        },
    }
}

pub async fn run_all_gates(iii: &III, gates: &[GateSpec], cwd: &Path) -> Vec<GateRunResult> {
    let mut out = Vec::with_capacity(gates.len());
    for gate in gates {
        let outcome = run_gate(iii, gate, cwd).await;
        out.push(GateRunResult {
            function_id: gate.function_id.clone(),
            description: gate.description.clone(),
            ok: outcome.ok,
            reason: outcome.reason,
        });
    }
    out
}

pub fn all_passed(results: &[GateRunResult]) -> bool {
    results.iter().all(|r| r.ok)
}
